packages/@lwc/engine-dom/src/apis/build-custom-element-constructor.ts

```diff
@@ -49,16 +50,26 @@ export function buildCustomElementConstructor(Ctor: ComponentConstructor): HTMLE
     const HtmlPrototype = getComponentHtmlPrototype(Ctor);
 
     return class extends HtmlPrototype {
+        isHydrated = false;
         constructor() {
             super();
-            createVM(this, Ctor, {
-                mode: 'open',
-                owner: null,
-                tagName: this.tagName,
-            });
+            if (this.isConnected) {
+                hydrateComponent(this, Ctor, {});
+                this.isHydrated = true;
+            } else {
+                createVM(this, Ctor, {
+                    mode: 'open',
+                    owner: null,
+                    tagName: this.tagName,
+                });
+            }
         }
         connectedCallback() {
-            connectRootElement(this);
+            if (this.isHydrated) {
```

Comment:
Again, a comment here would be helpful. I had to think for a bit before realizing that this is `true` when `customElements.define` is called and the un-upgraded element exists, so it tests `this.isHydrated = true` in the constructor, and then it's `true` here.